File: src/eQuantic.UI.Components/DateTimePicker.cs
using eQuantic.UI.Primitives;

namespace eQuantic.UI.Components;

/// <summary>
/// A moment: the date and the time of day, as the two controls that already answer each half.
/// <para>
/// It COMPOSES rather than reimplements, and that is the whole design. A combined picker that
/// grew its own grid and its own list would be a third keyboard to get right and a third place
/// for the date arithmetic to drift; here the calendar's arrows and the list's options are the
/// ones already tested, and this owns exactly one thing — that the two halves make one value.
/// </para>
/// <para>
/// A moment needs both halves before it is a moment, so nothing is reported until both are in.
/// Picking a date first and a time second is the order people work in, and either can be
/// re-picked afterwards without losing the other.
/// </para>
/// </summary>
public sealed class DateTimePicker : StatefulComponent
{
    private DateOnly? _date;
    private TimeOnly? _time;

    public DateTimePicker(DateTime? selected = null, Action<DateTime>? onChanged = null,
        DateTime? min = null, DateTime? max = null, int stepMinutes = 30,
        string dateLabel = "", string timeLabel = "")
    {
        Selected = selected;
        OnChanged = onChanged;
        Min = min;
        Max = max;
        StepMinutes = stepMinutes;
        DateLabel = dateLabel;
        TimeLabel = timeLabel;
        _date = selected is { } value ? DateOnly.FromDateTime(value) : null;
        _time = selected is { } moment ? TimeOnly.FromDateTime(moment) : null;
    }

    public DateTime? Selected { get; private set; }
    public Action<DateTime>? OnChanged { get; private set; }
    public DateTime? Min { get; private set; }
    public DateTime? Max { get; private set; }
    public int StepMinutes { get; private set; }
    public string DateLabel { get; private set; }
    public string TimeLabel { get; private set; }
    public bool Disabled { get; init; }

    public override void AdoptConfig(UiComponent next)
    {
        if (next is not DateTimePicker fresh) return;
        // NULL is a value the app can hand down, so the comparison includes it: a parent that
        // clears the moment is moving it as surely as one that changes it, and a picker still
        // showing yesterday's halves after a Clear button is the bug that reads as "it ignored me".
        var moved = fresh.Selected != Selected;
        Selected = fresh.Selected;
        OnChanged = fresh.OnChanged;
        Min = fresh.Min;
        Max = fresh.Max;
        StepMinutes = fresh.StepMinutes;
        DateLabel = fresh.DateLabel;
        TimeLabel = fresh.TimeLabel;
        // The app moved the moment, so both halves follow it — the same rule the Calendar keeps
        // for its view, and for the same reason: a control that ignores the value it was handed
        // looks broken long before anyone calls it wrong.
        if (moved)
        {
            _date = Selected is { } value ? DateOnly.FromDateTime(value) : null;
            _time = Selected is { } moment ? TimeOnly.FromDateTime(moment) : null;
        }
    }

    public override VisualNode Build(ComponentContext context)
    {
        var row = new Row(gap: Space.S3) { Cross = CrossAlign.Start, Width = SizeValue.Fill };
        row.Add(new Flexible(new DatePicker(_date, PickDate,
            Min is { } lower ? DateOnly.FromDateTime(lower) : null,
            Max is { } upper ? DateOnly.FromDateTime(upper) : null,
            DateLabel)
        { Disabled = Disabled }));
        row.Add(new Flexible(new TimePicker(_time, PickTime, StepMinutes, label: TimeLabel)
        { Disabled = Disabled }));
        return row;
    }

    private void PickDate(DateOnly day)
    {
        SetState(() => _date = day);
        Report();
    }

    private void PickTime(TimeOnly time)
    {
        SetState(() => _time = time);
        Report();
    }

    /// <summary>Both halves, or nothing: half a moment is not a value an app can store.</summary>
    private void Report()
    {
        if (_date is not { } day || _time is not { } time) return;
        var moment = day.ToDateTime(time);
        // The bounds are on the MOMENT, not on its halves — 17 July is inside a range that ends at
        // 17 July 09:00 only until the time says otherwise, which is exactly the case a
        // date-only check would wave through.
        if (Min is { } min && moment < min) return;
        if (Max is { } max && moment > max) return;
        Selected = moment;
        OnChanged?.Invoke(moment);
    }
}

File: src/eQuantic.UI.Runtime/src/shared/__transpiled__/DateTimePicker.ts
import { $eq, BuildContext, DateOnly, DatePicker, Flexible, Row, SharedStatefulComponent, SizeValue, TimeOnly, TimePicker, UiComponent } from "@equantic/runtime";

export class DateTimePicker extends SharedStatefulComponent {
    _date: any;
    _time: any;
    static $hydration = { _date: 'dateOnly', _time: 'timeOnly' };
    declare selected: any;
    declare onChanged: any;
    declare min: any;
    declare max: any;
    declare stepMinutes: number;
    declare dateLabel: string;
    declare timeLabel: string;
    declare disabled: boolean;

    constructor(selected: any = null, onChanged: any = null, min: any = null, max: any = null, stepMinutes: any = 30, dateLabel: any = '', timeLabel: any = '', props?: any) {
        super();
        if (selected !== undefined) this.selected = selected;
        if (onChanged !== undefined) this.onChanged = onChanged;
        if (min !== undefined) this.min = min;
        if (max !== undefined) this.max = max;
        if (stepMinutes !== undefined) this.stepMinutes = stepMinutes;
        if (dateLabel !== undefined) this.dateLabel = dateLabel;
        if (timeLabel !== undefined) this.timeLabel = timeLabel;
        if (this.stepMinutes === undefined) this.stepMinutes = 0;
        if (this.disabled === undefined) this.disabled = false;
        this.selected = selected;
        this.onChanged = onChanged;
        this.min = min;
        this.max = max;
        this.stepMinutes = stepMinutes;
        this.dateLabel = dateLabel;
        this.timeLabel = timeLabel;
        let value: any; 
        this._date = (value = selected) != null ? $eq.time.dateOnly.fromDateTime(value) : null;
        let moment: any; 
        this._time = (moment = selected) != null ? $eq.time.timeOnly.fromDateTime(moment) : null;
        if (props && typeof props === 'object') Object.assign(this, props);
    }

    build(_context: BuildContext) {
        let row = new Row(12, 'start', 'center', false, null, null, { cross: 'start', width: SizeValue.fill });
        let lower: any; let upper: any; 
        row.add(new Flexible(new DatePicker(this._date, this.pickDate.bind(this), (lower = this.min) != null ? $eq.time.dateOnly.fromDateTime(lower) : null, (upper = this.max) != null ? $eq.time.dateOnly.fromDateTime(upper) : null, this.dateLabel, { disabled: this.disabled })));
        row.add(new Flexible(new TimePicker(this._time, this.pickTime.bind(this), this.stepMinutes, null, null, this.timeLabel, { disabled: this.disabled })));
        return row;
    }

    adoptConfig(next: UiComponent) {
        let fresh: any; 
        if (!((next instanceof DateTimePicker && (fresh = next, true)))) return;
        let moved = !$eq.equals(fresh.selected, this.selected);
        this.selected = fresh.selected;
        this.onChanged = fresh.onChanged;
        this.min = fresh.min;
        this.max = fresh.max;
        this.stepMinutes = fresh.stepMinutes;
        this.dateLabel = fresh.dateLabel;
        this.timeLabel = fresh.timeLabel;
        if (moved) {
            let value: any; 
            this._date = (value = this.selected) != null ? $eq.time.dateOnly.fromDateTime(value) : null;
            let moment: any; 
            this._time = (moment = this.selected) != null ? $eq.time.timeOnly.fromDateTime(moment) : null;
        }
    }

    pickDate(day: DateOnly) {
        this.setState(() => this._date = day);
        this.report();
    }

    pickTime(time: TimeOnly) {
        this.setState(() => this._time = time);
        this.report();
    }

    report() {
        let day: any; let time: any; 
        if (!((day = this._date) != null) || !((time = this._time) != null)) return;
        let moment = day.toDateTime(time);
        let min: any; 
        if ((min = this.min) != null && (moment.compareTo(min) < 0)) return;
        let max: any; 
        if ((max = this.max) != null && (moment.compareTo(max) > 0)) return;
        this.selected = moment;
        this.onChanged?.(moment);
    }
}

